Propagate converted fields through projections

MattAlp opened the pull request:
Addresses #137187.

Reviewers should check

- Transformation rule correctness
- Comprehensiveness of the analyzer test
- If we should add something to the logical plan optimization tests

MattAlp committed:
Propagate synthetic attributes directly through projections

MattAlp committed:
WIP

MattAlp committed:
Copy/paste error nit

MattAlp committed:
Catch other copy/paste issue

MattAlp committed:
WIP projection test

MattAlp committed:
Rewrite synthetic attribute projection test

MattAlp committed:
Update docs/changelog/137923.yaml

MattAlp committed:
Merge branch 'main' into propagate-converted-fields-thru-projections

fang-xing-esql reviewed (commented):
Thank you @MattAlp ! Can we add some `LogicalPlanOptimizerTests` and `CsvTests` to validate this change? The original issue errors out in logical plan's verifier, it will be helpful to validate the plan after logical planner is as expected and we get expected results. I tried this query below, it failed on main, and run to completion on the branch, however it seems like we get an extra field returned.

```
+ curl -u elastic:password -v -H 'Content-Type: application/json' '127.0.0.1:9200/_query?format=txt' -d '{
    "query": "from sample_data_* | keep @timestamp, message | mv_expand @timestamp | eval message = message::keyword"
}
'
       @timestamp       |$$message$converted_to$keyword|       message       
------------------------+------------------------------+---------------------
2023-10-23T12:15:03.360Z|Connected to 10.1.0.3         |Connected to 10.1.0.3
2023-10-23T12:27:28.948Z|Connected to 10.1.0.2         |Connected to 10.1.0.2
2023-10-23T13:33:34.937Z|Disconnected                  |Disconnected         
2023-10-23T13:51:54.732Z|Connection error              |Connection error     
2023-10-23T13:52:55.015Z|Connection error              |Connection error     
2023-10-23T13:53:55.832Z|Connection error              |Connection error     
2023-10-23T13:55:01.543Z|Connected to 10.1.0.1         |Connected to 10.1.0.1
2023-10-23T12:15:03.360Z|Connected to 10.1.0.3         |Connected to 10.1.0.3
2023-10-23T12:27:28.948Z|Connected to 10.1.0.2         |Connected to 10.1.0.2
2023-10-23T13:33:34.937Z|Disconnected                  |Disconnected         
2023-10-23T13:51:54.732Z|Connection error              |Connection error     
2023-10-23T13:52:55.015Z|Connection error              |Connection error     
2023-10-23T13:53:55.832Z|Connection error              |Connection error     
2023-10-23T13:55:01.543Z|Connected to 10.1.0.1         |Connected to 10.1.0.1
```

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1952,6 +1952,23 @@ private static LogicalPlan addGeneratedFieldsToEsRelations(LogicalPlan plan, Lis
                 }
                 return esr;
             });
+            res = res.transformUp(Project.class, p -> {
```
It seems that just adding the missing synthetic attribute won't be enough.
What happens is that the `Project` still contains `FieldAttribute`(s) containing an `InvalidMappedField` that the `UnionTypesCleanup` rule converts to an `UnsupportedAttribute`. Because of this, the resulting plan is not `resolved()`. Which means that the rule [will not add](https://github.com/elastic/elasticsearch/blob/9426456b4af91a0e6f4a8c7972e1f9ca82dbe96f/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java#L2125) a top `Project` to drop the synthetic attributes. So we end up outputting more attributes than we need to.

I think we will need to:
- [either] not only add the synthetic attributes to the Project (as done so far), but drop the _original_ attribute corresponding to to the synthetic one. This should be safe, since no other command is allowed to work with this attribute before conversion; but I'm not sure if the verification of the plan will remain safe then and we'll put out the correct error message to the user, in case the query does touch the attribute before conversion; an alternative would be to:
- [or] add a `Project` before the converting `EVAL` that drops the original attribute corresponding to to the synthetic one.

alex-spies replied:
> Because of this, the resulting plan is not resolved()

Wow, that'd be horrific! But actually, it's not the case, but it's rightfully confusing. The only plan where an UnsupportedAttribute can occur in the expressions should be EsqlProject; and that one specifically considers UnsupportedAttributes okay: https://github.com/elastic/elasticsearch/blob/3d0a64741a9ff21eb5d6bb5276077e0ddc4c5840/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/local/EsqlProject.java#L74

bpintea reviewed (commented):
Once we get the right solution, we should add integration tests too: adding to `union-types.csv-spec` might be the easiest (there are also yaml ones in `160_union_types.yml`).

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1952,6 +1952,23 @@ private static LogicalPlan addGeneratedFieldsToEsRelations(LogicalPlan plan, Lis
+
+                List<NamedExpression> newProjections = new ArrayList<>(p.projections());
+                newProjections.addAll(syntheticAttributesToCarryOver);
+                return new Project(p.source(), p.child(), newProjections);
```
++, approach looks correct to me!

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:
```diff
@@ -4431,6 +4432,37 @@ public void testBucketWithIntervalInStringInGroupingReferencedInAggregation() {
+        IndexResolution resolution = IndexResolution.valid(index);
+        Analyzer analyzer = analyzer(resolution);
+
+        String query = "FROM union_index_combined | KEEP id, foo | MV_EXPAND foo | EVAL id = id::keyword";
```
note: the analyzer doesn't push down projections, yet, so the MV_EXPAND shouldn't be required for making a test.

alex-spies reviewed (approved):
Solution looks correct to me, thanks @MattAlp !

You got some suggestions on added tests - I agree these make sense adding before merging!

@bpintea 's comment [here](https://github.com/elastic/elasticsearch/pull/137923#discussion_r2520319879) is important. I'm quite convinced that we're fine, but that's a great rabbit hole to dig into to understand why we're, indeed, fine. It'd be cool if you could have a look at this, and add some comments in places that would make this easier to understand for our future selves. Also, there are existing csv tests for union types that ensure that the synthetic attributes don't slip through - maybe have a look at them if you want to run the debugger and confirm what's happening.

What I can say so far is:
- plans must be resolved after passing the verifier, otherwise we cannot e.g. call `.datatype()` on expressions and that blows up the optimizer.
- UnsupportedAttribute is actually Unresolvable. Generally, that means that plans and expressions containing an UnsupportedAttribute _will_ be consider unresolved.
This makes it so that using it in expressions will generally trigger an error message from the verifier; that's what we want.
- Using UnsupportedAttributes is okay in EsqlProject, and the verifier has an exception to allow it. That's because it's fine to use a field with mapping conflicts in KEEP and DROP (but not elsewhere, and not in RENAME).

MattAlp committed:
Update x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java

Co-authored-by: Bogdan Pintea <sig11@mailbox.org>

MattAlp committed:
Update x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java

Co-authored-by: Bogdan Pintea <sig11@mailbox.org>

MattAlp committed:
Update analyzer tests to clean up projection of unsupported expressions post-cast

MattAlp committed:
Added testUnionTypesResolvePastProjections to logical plan optimizer tests

MattAlp committed:
Create CSV tests for union type casting thru project

MattAlp committed:
Correct the retention & rejection of synthetic & unresolved attributes

Keep synthetics if they aren't aliased, keep unsupported attributed if they're explicitly retained, drop either if shadowing (with optional casting) occurs

MattAlp committed:
Add tests for retention of unsupported & simultaneously-cast fields

MattAlp committed:
Merge remote-tracking branch 'origin/main' into propagate-converted-fields-thru-projections

MattAlp committed:
Update EsIndex constructors and parser method after merge

Minor divergence due to ~1mo of staleness

MattAlp committed:
Test-generated diff for inline cast

MattAlp commented:
I haven't added many comments inline as part of this PR - I'd actually like to add them to the internal & external docs alongside some other notes I've been taking while working on planner PRs incl. this one.

julian-elastic commented on docs/reference/query-languages/esql/kibana/definition/inline_cast.json:
```diff
@@ -7,6 +7,7 @@
   "date" : "to_datetime",
   "date_nanos" : "to_date_nanos",
   "date_period" : "to_dateperiod",
+  "date_range" : "to_date_range",
```
Why is this needed?

MattAlp replied:
This might be a potential smell relating to code/doc-gen, but I did not investigate it

MattAlp commented:
`INLINE STATS` behavior is covered - need to explore `STATS` further

julian-elastic reviewed (approved):
Overall looks good, thank you for taking the time to explain the fix to me! I left a few comments about additional tests to make sure we have handled all cases.

MattAlp commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -9844,6 +9846,97 @@ public void testFullTextFunctionOnEvalNull() {
     }
 
+    /**
+     * Project[[foo{r}#12, $$id$converted_to$keyword{f$}#13 AS id#9]]
```
See also https://github.com/elastic/elasticsearch/pull/136398#pullrequestreview-3464640810. Alex is back after Tuesday, merging will wait until he weighs in.

julian-elastic replied:
@alex-spies Do you agree the !id should be pruned from the project? What rule is responsible for pruning attributes no longer needed from Projects? I looked at PruneColumns(), but it seems to do limited pruning for Projects.

MattAlp committed:
Merge branch 'main' into propagate-converted-fields-thru-projections

MattAlp committed:
Merge branch 'main' into propagate-converted-fields-thru-projections

alex-spies reviewed (approved):
Nice work @MattAlp !

MattAlp committed:
Make Project transformation conditional within ResolveUnionTypes

MattAlp committed:
Merge branch 'main' into propagate-converted-fields-thru-projections
